tox.ini

```diff
@@ -0,0 +1,100 @@
+[tox]
+setupdir = {toxinidir}
+envlist =
+    unittests-{py37,py38,py39,py310,py311},
+
+    aiohttp-{py37,py38,py39,py310,py311}-aiohttplatest,
+    chalice-{py37,py38,py39,py310,py311}-chalicelatest,
+    channels-{py37,py38,py39,py310,py311}-channelslatest,
+    django-{py37,py38,py39,py310,py311}-djangolatest,
+    django-py310-django{0400,0302},
+    fastapi-{py37,py38,py39,py310,py311}-fastapilatest,
+    flask-{py37,py38,py39,py310,py311}-flasklatest,
+    mypy-{py37,py38,py39,py310,py311}-mypylatest,
+    pyright-{py37,py38,py39,py310,py311}-pyrightlatest,
+    sanic-{py37,py38,py39,py310,py311}-saniclatest,
+    starlette-{py37,py38,py39,py310,py311}-starlettelatest,
+    starlette-py310-starlette{001701,001800,001901,002004}
+
+[testenv]
+deps =
+    pytest
+    pytest-asyncio
+    pytest-benchmark
+    pytest-cov
+    pytest-emoji
+    pytest-mock
+    pytest-mypy-plugins
+    pytest-snapshot
+    mypy
+
+    unittests: asgiref
+    unittests: freezegun
+    unittests: opentelemetry-api<2
+    unittests: opentelemetry-sdk<2
+    unittests: pydantic[email]
+    unittests: pygments
+    unittests: pyright
+    unittests: types-aiofiles
+    unittests: types-certifi
+    unittests: types-chardet
+    unittests: types-freezegun
+    unittests: types-python-dateutil
+    unittests: types-requests
+    unittests: types-setuptools
+    unittests: types-toml
+    unittests: types-typed-ast
+    unittests: types-ujson
+
+    chalice: werkzeug
+    chalice-chalicelatest: chalice
+    channels: asgiref
+    channels: pytest-django
+    channels-channelslatest: channels
+    django: asgiref
+    django: pytest-django
+    django-djangolatest: Django
+    django-django0400: Django<4.1
+    django-django0302: Django<3.3
+    flask: pytest-flask
+    flask-flasklatest: flask[async]
+    pyright-pyrightlatest: pyright
+    aiohttp: pytest-aiohttp
+    aiohttp-aiohttplatest: aiohttp
+    fastapi: python-multipart
+    fastapi: requests
+    fastapi-fastapilatest: fastapi
+    mypy: pydantic
+    mypy-mypylatest: mypy
+    sanic: sanic-testing>=22.3,<22.4
+    sanic-saniclatest: sanic>=20.12.2,<22.0.0
+    starlette: click
+    starlette: httpx
+    starlette: pygments
+    starlette: python-multipart
+    starlette: requests
+    starlette: uvicorn
+    starlette-starlettelatest: starlette
+    starlette-starlette001701: starlette==0.17.1
+    starlette-starlette001800: starlette==0.18.0
+    starlette-starlette001901: starlette==0.19.1
+    starlette-starlette002004: starlette==0.20.4
```

Comment:
this feels a lot more complicate than what it should be if I'm honest. 

I do get the benefits of not using a lock file, but I don't really like the idea of having to duplicate the dependencies here

Reply:
You CAN use poetry somehow if you really want to manage these, but the point of using a test dispatcher like tox is to fully separate your concerns. It's much cleaner to specify exactly what dependencies are needed for testing with different suites by using a separate service.

If you have poetry install every dependency as a base and then just have tox manage the version matrixed dependencies that's not terrible, but it also sort of misses the point that having other dependencies can causes crashes and dependency hell when you try to reinstall with an older version of some package, if it shares a dependency with another package that's not under test but was included in the base poetry config.

A good example of problematic side effects would be FastAPI. They're frequently trailing starlette releases by a good margin. If you install everything with poetry first, FastAPI will likely cause starlette tests to be run with older versions of starlette and not the latest, whether or not you want it to be. That silent and unintended side effect would prevent you from testing on newer releases of starlette until FastAPI updates themselves. For libraries that are trying to guarantee compatibility with up to date versions of all their integrations that can easily hide critical failures for months. To get the most consistent and trustworthy results you want to test with only the dependencies you absolutely need for each integration.

Personally I don't think poetry is up to the task at all of managing any test dependencies of libraries. It's really good as a tool for managing the environment of applications, and for getting libraries published with minimal config. But if you're building a library and not an application, I don't think it's a very good tool for managing tests.

Reply:
And actually thinking back, that was the exact problem with your testing environment that stalled #1594 in the first place prompting me to open an issue about this. If you use poetry to continue managing your base environment and include all your other dependencies in there like starlette and FastAPI you'll be only marginally more resilient and miss the original root cause.

Reply:
> You CAN use poetry somehow if you really want to manage these, but the point of using a test dispatcher like tox is to fully separate your concerns. It's much cleaner to specify exactly what dependencies are needed for testing with different suites by using a separate service.

would recommend removing the dev dependencies and always use tox then?

> Personally I don't think poetry is up to the task at all of managing any test dependencies of libraries. It's really good as a tool for managing the environment of applications, and for getting libraries published with minimal config. But if you're building a library and not an application, I don't think it's a very good tool for managing tests.

Ok, I think I'm getting convinced 😊 is it ok if I revert the test for the time being and use markers for dividing them? I don't know when I'll merge that other PR, and I want to keep the tests where they are for now (there's also quite a few PRs in flight)

Reply:
That's fine, I would definitely recommend eventually going the 1 test suite per folder route but if it's easier you can always make that change later. Feel free to edit this PR in place and use it. I don't really have much more time to invest in actually writing code for at least a week.

Reply:
> would recommend removing the dev dependencies and always use tox then?

I personally just use tox but it certainly takes a good 20-30 seconds sometimes to set up a venv before each test run. I have a bash alias for sourcing the tox env's created virtual environment after 1 run and editable installing the package into it. That being said, it's mostly because of how obnoxiously many tox environments we have (I think 500+). 

That being said, it might be easier for development workflows to have all those dev dependencies and to just use pytest directly. You can rely on your CI to handle the matrix testing and ensure nothing slips by, and focus on development with just any singular version of the framework you're testing against.

So really it just comes down to personal preference for local development. I can see the value either way.